src/chart-transforms/encoding.js

```diff
@@ -245,6 +245,22 @@ export const SelectEncodingField = ( { setAttributes, json, field, label } ) =>
 		}
 	}, [ field, json, setAttributes ] );
 
+	const encodingField = getEncodingField( json, field );
+
+	// If the selected field is not available in the field options, pick one
+	// that is valid. This means that changing dataset will clear previous
+	// field settings in your chart.
+	useEffect( () => {
+		if ( ! fieldOptions.length || ! fieldOptions[0].value ) {
+			return;
+		}
+
+		const selectedOptionExists = ! ! fieldOptions.find( ( { value } ) => encodingField === value );
+		if ( ! selectedOptionExists ) {
+			onChangeField( fieldOptions[0].value );
+		}
+	}, [ encodingField, fieldOptions, onChangeField ] );
```

Comment:
Micro-optimization: since `onChangeField` already declares `fieldOptions` as a dependency, it can be removed here... since any change to `fieldOptions` will also update `onChangeField`. (unless that breaks linting, in which case never mind...)

Reply:
Yeah, it breaks linting 😑